### PR TITLE
Add support for lockfile in Gemfile and bundle install --no-lock

### PR DESCRIPTION
**Pull request by jeremyevans**
## What was the end-user or developer problem that led to this PR?

I would like a way to install dependencies from a Gemfile without creating a lockfile. I would also like a way to mark a library Gemfile as not creating lockfiles, so that users who use the Gemfile to install the development dependencies are not locked to those specific dependency versions going forward (the library should work with the entire range of libraries mentioned in the Gemfile).

Other developers wanted the ability to store the lock file at a non-default location, so this adds support for both features.

## What is your fix for the problem, implemented in this PR?

This adds support for `lockfile` in Gemfile and `bundle install --no-lock`

When `lockfile "path"` is in a Gemfile, bundle install will create a lock file
at the given path. This is useful for situations where the lock file will
differ per Ruby platform, where the directory containing Gemfile is
read only, and any other situation where you want the lock file in
a non-default location.

You can use `lockfile false` to prevent writing the lock file.
This is useful for library Gemfiles or any other situation where
you would like to be able to easily install dependencies from a
Gemfile without being locked to those versions in the future.

`bundle install --no-lock` allows for the same behavior as including
`lockfile false` in the Gemfile. This allows you to get the behavior
without modifying the Gemfile, which is useful if you do not control
the Gemfile.

This adds support for `BUNDLE_GEMFILE_LOCK` environment variable
for specifying the lock file location. This is useful for `bundle exec`, or
when you don't want to modify a Gemfile to include `lockfile`, but still
want the lock file stored in a non-default location.

This change is backwards compatible. The behavior only changes if
you are using the new features.

## Make sure the following tasks are checked

- [X] Describe the problem / feature
- [X] Write [tests](https://github.com/ruby/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [X] Write code to solve the problem
- [X] Make sure you follow the [current code style](https://github.com/ruby/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/ruby/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by jeremyevans** [comment]
@byroot when you have time, could I please get your feedback on this?

**Comment by byroot** [comment]
> could I please get your feedback on this?

Honestly I don't really see the upside over adding `Gemfile.lock` in `.gitignore`, and occasionally `rm Gemfile.lock`, but why not I guess?

File: bundler/lib/bundler/man/gemfile.5.ronn [hunk 1]
```diff
@@ -556,3 +556,13 @@ bundler uses the following priority order:
...
      a warning and will abort with an error in the future.
+
+## NO LOCK
+
+By default, Bundler will create a lock file when installing gems. If you do not
+wish to create a lock file, use `no_lock!` in the Gemfile:
+
+    no_lock!
+
+This is useful for library development and other situations where the code is
```
Comment (colby-swandale) [inline bundler/lib/bundler/man/gemfile.5.ronn:574]: Could you shed a light on what these other situations/use cases are? Are they also related to gem maintainership?

**Comment by kou** [comment]
It may not be related but I sometimes want to generate `Gemfile.lock` in different directory. It's like an out-of-source build in C. For example, https://github.com/ruby-gnome/ruby-gnome/ uses Docker to test in many environments. Docker container and host may use different uid. So I don't want to change the source directory. (I don't want to create `Gemfile.lock` in source directory.)

If the feature is implemented, `BUNDLE_GEMFILE_LOCK=/dev/null bundle install` may also solve this situation.

**Comment by Earlopain** [comment]
Disclaimer: I did not test this out, this is based on my understanding on how it would work. Please correct me if anything is wrong.

This causes bundler to re-resolve every time I guess? Downloading the index, refetching git gems, etc. This adds quite a few seconds to each bundler command (depending on your internet connection). Also, no offline development.

I guess it's fine as an opt-in for each user, but I would not want that to be decided for me with `no_lock!`.

Speaking of git sources, can you `bundle exec foo` with this? Currently it looks like bundler _needs_ a lockfile for that to work.

>  are not locked to those specific dependency versions going forward (the library should work with the entire range of libraries mentioned in the Gemfile).

This doesn't really make sense to me. You can install and use any version in your lockfile that your constraints allow, you are just given one specific version (likely the latest at the time of resolving) unless you yourself change/update that. When the lockfile is gone, wouldn't you always get the most recent version? So you end up with less testing of older versions.

File: bundler/lib/bundler/man/gemfile.5.ronn (hunk 1 above)
Comment (jeremyevans) [reply bundler/lib/bundler/man/gemfile.5.ronn:574]: Any case where you are developing libraries that are designed to work with multiple versions of dependencies (that is, almost every library). Otherwise, as @byroot alluded to above:

* Unless you add `Gemfile.lock` to `.gitignore`, it will show up as an untracked file, which is generally undesirable in terms of development.
* Unless you `rm Gemfile.lock`, you are likely stuck testing the library with the dependencies that were present when the libraries were initally installed.

Other issues:

* If you want to test your library with multiple Ruby versions, `Gemfile.lock` gets in the way as it restricts the Ruby version. 
* Creating the `Gemfile.lock` does not work in cases where the process doesn't have write access to the directory containing `Gemfile`.

In general, `Gemfile.lock` is only useful for applications. There is rarely a good reason to have it for a library, and it generally just gets in the way.

This is only related to gem maintainership in that the vast majority of gems are libraries and would benefit from this.

Hopefully, these reasons make sense to you. If not, can you explain what benefit you see from creating a `Gemfile.lock` for a library `Gemfile`? Additionally, even if you believe some libraries may benefit from a `Gemfile.lock`, can you at least agree that not all libraries may want a `Gemfile.lock`? This isn't a default behavior, it's purely opt-in, so it causes no problems for users who want to continue having a `Gemfile.lock` created.

FWIW, the `--no-lock` option is already supported for `gem` (`gem install -g Gemfile --no-lock`).

**Comment by jeremyevans** [comment]
> Disclaimer: I did not test this out, this is based on my understanding on how it would work. Please correct me if anything is wrong.

You can test out how this will work by including this in the `Gemfile`:

```
::Bundler::Definition.no_lock = true
```

This just allows it to be a local setting without changing a global setting in Bundler. That support was added back in 2020.

> This causes bundler to re-resolve every time I guess? Downloading the index, refetching git gems, etc. This adds quite a few seconds to each bundler command (depending on your internet connection). Also, no offline development.

In general, `bundle exec` works with this, as well as `Bundler.setup` and `Bundler.require`. Those already support cases where `Gemfile` exists and `Gemfile.lock` does not. This just makes it so `Gemfile.lock` is not written.

> I guess it's fine as an opt-in for each user, but I would not want that to be decided for me with `no_lock!`.

This is already possible via `::Bundler::Definition.no_lock = true`, as mentioned above. This just gives a simpler method, and avoids making a global setting change.

> Speaking of git sources, can you `bundle exec foo` with this? Currently it looks like bundler _needs_ a lockfile for that to work.

From my testing, git sources do require a lockfile currently for `bundle exec` to work. This is unlikely to be a major issue for a library `Gemfile`, since git sources are not that common in that situation (unlike application `Gemfile`s).

> > are not locked to those specific dependency versions going forward (the library should work with the entire range of libraries mentioned in the Gemfile).
> 
> This doesn't really make sense to me. You can install and use any version in your lockfile that your constraints allow, you are just given one specific version (likely the latest at the time of resolving) unless you yourself change/update that. When the lockfile is gone, wouldn't you always get the most recent version? So you end up with less testing of older versions.

In general, you get the latest already installed version that meets the version requirements. `bundle exec` doesn't download new versions.

IMO, for normal development, it's generally better to test with newer installed dependency versions than always testing the dependency version you originally used when you installed the library.  Testing older versions of libraries is good, but that tends to be a CI type situation, or a case where a library would use multiple `Gemfile`s with different dependency versions (e.g. rack-test's `Gemfile` and `Gemfile.rack-1.x`).

File: bundler/lib/bundler/man/gemfile.5.ronn (hunk 1 above)
Comment (byroot) [reply bundler/lib/bundler/man/gemfile.5.ronn:574]: > There is rarely a good reason to have it for a library, and it generally just gets in the way.

Well it depends. As @Earlopain mentioned, it's important for performance. e..g on `ruby/json`, the Gemfile.lock is git ignored, so every time I need to test with JRuby, I got to re-do the resolution:

```
real	0m16.573s
user	0m42.594s
sys	0m2.733s
```

And that's a project with basically no dependencies.

> If you want to test your library with multiple Ruby versions, Gemfile.lock gets in the way as it restricts the Ruby version.

It shouldn't. The only issue is with ruby-head and precompiled binaries, I complained a lot about that, and it was somewhat improved lately.

> If not, can you explain what benefit you see from creating a Gemfile.lock for a library Gemfile

I maintain projects with both approachs. E.g. `ruby/json` ignores the Gemfile.lock, but `redis-client` commits it and use dependabot to have regular PRs to bump dependencies. It is actually an OK workflow, as I'm notified when a dependency release a new version, and immediately get a CI run which proves or disprove compatibilty with it.

But it's OK because `redis-client` has few dependencies. With e.g. Rails, the Gemfile is much bigger and that workflow would be a PITA.

Overall, I think the more useful feature would be a way to set the path for `Gemfile.lock` in the `Gemfile`, e.g.

```ruby
lock_file RUBY_PLATFORM =~ /java/ ? "Gemfile.jruby.lock" : "Gemfile.lock"
```

And it could allow `lock_file false` for your use case.

Comment (Earlopain) [reply bundler/lib/bundler/man/gemfile.5.ronn:574]: https://github.com/ruby/rubygems/pull/9059#issuecomment-3510323291

> You can test out how this will work by including this in the Gemfile

Ah, I didn't realize. Although, doing that still generates the lockfile for me, doesn't really want to work.

Personally, I don't really see the use case though. If you don't commit your lockfile, CI will test latest gem versions for you. If you do commit it, you get deterministic versions anyways. If you still have some old gem version in your lockfile, that shouldn't matter since your gem still has to work with it (since there are no constraints against it).

The only I can think of is ruby version bumps (which includes ruby-dev for me), I very rarely remove the lockfile to get rid of warnings that new ruby versions introduce. But to me that is fine for me, paying the cost to re-resolve with every command is just not worth it.

I would not take JRuby as a good counter-example for this, it is already consistently the slowest in CI (occasionally I use it locally where it is more of the same). But like I said earlier, even on CRuby it takes a few seconds (2.3s on the `rubocop` repo, subtracting the overhead of `bundle lock` when the lockfile already exists)

Comment (jeremyevans) [reply bundler/lib/bundler/man/gemfile.5.ronn:574]: > Overall, I think the more useful feature would be a way to set the path for `Gemfile.lock` in the `Gemfile`, e.g.
> 
> ```ruby
> lock_file RUBY_PLATFORM =~ /java/ ? "Gemfile.jruby.lock" : "Gemfile.lock"
> ```
> 
> And it could allow `lock_file false` for your use case.

Since @kou also thought that would be a useful feature, and it allows support for a superset of my need, I'll work on support for that.

Comment (kou) [reply bundler/lib/bundler/man/gemfile.5.ronn:574]: Thanks!

Comment (jeremyevans) [reply bundler/lib/bundler/man/gemfile.5.ronn:574]: @byroot @kou I've updated this to use `lock_file` instead of `no_lock!` in the `Gemfile`, and added support for the `BUNDLE_GEMFILE_LOCK` environment variable.

Comment (byroot) [reply bundler/lib/bundler/man/gemfile.5.ronn:574]: Thank you! I’ll try to find some time tonight to play with it, to see how it does

File: bundler/spec/install/gemfile_spec.rb [hunk 2]
```diff
@@ -27,6 +27,20 @@
...
+        source "https://gem.repo1"
+        gem 'myrack'
+      G
+
+      bundle :install, gemfile: bundled_app("NotGemfile")
+
+      ENV["BUNDLE_GEMFILE"] = "NotGemfile"
+      ENV["BUNDLE_GEMFILE_LOCK"] = "ReallyNotGemfile.lock"
+      expect(the_bundle).to include_gems "myrack 1.0.0"
+    end
```
Comment (kou) [inline bundler/spec/install/gemfile_spec.rb]: It seems that we need to split this to `lock_file` only case and `BUNGLE_GEMFILE_LOCK` case. If we have both of them and they use the same value, we can't test `lock_file` and `BUNDLE_GEMFILE_LOCK`.

How about splitting this to `lock_file` only case and `BUNDLE_GEMFILE_LOCK` only case`?

Comment (jeremyevans) [reply bundler/spec/install/gemfile_spec.rb]: We have a test for the `lock_file` only case in `bundler/spec/commands/install_spec.rb‎`. I'm not sure whether it would make sense to test the `BUNDLE_GEMFILE_LOCK` only case. If you aren't using `lock_file` in the `Gemfile`, `bundler` is going to append `.lock` to create the lock file path. I think it only makes sense to use `BUNDLE_GEMFILE_LOCK` if you are also using `lock_file` to change the default lock file path, unless you are moving the location of the lock file after it is created. We could have a test doesn't use `lock_file`,  and renames the lock files afterward, and then test that the renamed file works with `BUNDLE_GEMFILE_LOCK`. Is that what you'd like to see?

Comment (kou) [reply bundler/spec/install/gemfile_spec.rb]: Oh, I was going to use only `BUNDLE_GEMFILE_LOCK` case in Ruby-GNOME. (No `lock_file` in `Gemfile`.)

I don't want to specify `Gemfile.lock` path in `Gemfile` explicitly. (e.g.: `lock_file "/tmp/Gemfile.lock"` and `lock_file false`)
I'll specify `Gemfile.lock` by `BUNDLE_GEMFILE_LOCK` in Docker container something like:

```shell
mkdir -p /build
export BUNDLE_GEMFILE=/source/Gemfile
export BUNDLE_GEMFILE_LOCK=/build/Gemfile.lock
cd /build
bundle install # This must not change /source/
bundle exec ruby /source/execonf.rb # This must not change /source/
make -j$(nproc) # This must not change /source/
bundle exec ruby /source/run-test.rb # This must not change /source/
```

So how about the following case?

```ruby
      gemfile bundled_app("NotGemfile"), <<-G
        source "https://gem.repo1"
        gem 'myrack'
      G

      ENV["BUNDLE_GEMFILE_LOCK"] = "ReallyNotGemfile.lock"
      bundle :install, gemfile: bundled_app("NotGemfile")

      ENV["BUNDLE_GEMFILE"] = "NotGemfile"
      expect(the_bundle).to include_gems "myrack 1.0.0"
```

**Comment by byroot** [comment]
So I tested this branch by setting:

```ruby
lock_file "Gemfile.#{RUBY_ENGINE}.lock"
```

in `ruby/json` Gemfile, and it worked as I expected. Each ruby implementation had its own Gemfile.lock, which solved many annoyances for me. So huge 👍.

Note however that I'm not super familiar with bundler's implementation, so this is just me saying that this PR as it stand is a very useful feature, it says nothing about the quality of the patch of any possible corner cases etc.

File: bundler/spec/install/gemfile_spec.rb (hunk 2 above)
Comment (jeremyevans) [reply bundler/spec/install/gemfile_spec.rb]: Understood, thank you for explaining. I added a spec to test that `BUNDLE_GEMFILE_LOCK` correctly sets the lock file to create, if `lock_file` is not used in `Gemfile`.

File: bundler/spec/install/gemfile_spec.rb [hunk 3]
```diff
@@ -27,6 +27,35 @@
...
+    it "respects BUNDLE_GEMFILE_LOCK during bundle install" do
+      ENV["BUNDLE_GEMFILE_LOCK"] = "ReallyNotGemfile.lock"
+
+      gemfile bundled_app("NotGemfile"), <<-G
+        source "https://gem.repo1"
+        gem 'myrack'
+      G
+
+      bundle :install, gemfile: bundled_app("NotGemfile")
+      expect(File.exist?(bundled_app("ReallyNotGemfile.lock"))).to eq true
```
Comment (kou) [inline bundler/spec/install/gemfile_spec.rb]: Can we use this?

```suggestion
      expect(bundled_app("ReallyNotGemfile.lock")).to exist
```

**Review by kou — APPROVED** [review]
+1

File: bundler/spec/install/gemfile_spec.rb (hunk 3 above)
Comment (jeremyevans) [reply bundler/spec/install/gemfile_spec.rb]: That looks better to me. I merged it into the PR.

File: bundler/lib/bundler/dsl.rb [hunk 4]
```diff
@@ -101,6 +103,10 @@ def gem(name, *args)
       add_dependency(name, version, options)
     end
 
+    def lock_file(file)
```
Comment (hsbt) [inline bundler/lib/bundler/dsl.rb]: How about to use `lockfile` instead of `lock_file`? I used the terminalogy for that with `lockfile`, not "lock file" in RubyGems and Bundler.

File: bundler/lib/bundler/man/bundle-config.1.ronn [hunk 5]
```diff
@@ -176,6 +176,9 @@ learn more about their operation in [bundle install(1)](bundle-install.1.html).
    relative paths in the `Gemfile`, among other things. By default, bundler
    will search up from the current working directory until it finds a
    `Gemfile`.
+* `gemfile_lock` (`BUNDLE_GEMFILE_LOCK`):
```
Comment (hsbt) [inline bundler/lib/bundler/man/bundle-config.1.ronn]: I prefer to use `BUNDLE_LOCKFILE` instead of `BUNDLE_GEMFILE_LOCK` with same reason of https://github.com/ruby/rubygems/pull/9059/files#r2520523557

**Comment by jeremyevans** [comment]
> I'm +1 to add these features. I added to comment about naming. But the current proposal is okay to me.

Great! I'll work on renaming things.

**Comment by jeremyevans** [comment]
I've pushed changes to rename `lock_file` to `lockfile` and `BUNDLE_GEMFILE_LOCK` to `BUNDLE_LOCKFILE`.

File: bundler/lib/bundler/dsl.rb (hunk 4 above)
Comment (colby-swandale) [reply bundler/lib/bundler/dsl.rb]: This change looks like it's been added, will mark this as resolved.

File: bundler/lib/bundler/man/bundle-config.1.ronn (hunk 5 above)
Comment (colby-swandale) [reply bundler/lib/bundler/man/bundle-config.1.ronn]: This change looks like it's been added, will mark this as resolved.

File: bundler/lib/bundler/man/bundle-config.1.ronn [hunk 6]
```diff
@@ -176,6 +176,9 @@ learn more about their operation in [bundle install(1)](bundle-install.1.html).
    relative paths in the `Gemfile`, among other things. By default, bundler
    will search up from the current working directory until it finds a
    `Gemfile`.
+* `gemfile_lock` (`BUNDLE_LOCKFILE`):
```
Comment (colby-swandale) [inline bundler/lib/bundler/man/bundle-config.1.ronn]: We're documenting the `gemfile_lock` configuration option, but I don't see this option defined in the Bundler Settings config. I think this means setting `bundle config gemfile_lock <file>` won't work. Was this an intentional omission?

https://github.com/ruby/rubygems/blob/40321a746ec531c764396b4a302afa0e8fc28bf0/bundler/lib/bundler/settings.rb#L58-L74

File: bundler/lib/bundler/man/bundle-install.1.ronn
Comment (colby-swandale) [inline bundler/lib/bundler/man/bundle-install.1.ronn:13]: Could we add an entry for `--no-lock` into the SYNOPSIS ?

File: bundler/lib/bundler/man/bundle-config.1.ronn [hunk 7]
```diff
@@ -176,6 +176,9 @@ learn more about their operation in [bundle install(1)](bundle-install.1.html).
...
+   The name of the file that bundler should use as the `Gemfile.lock`. By default,
+   will add `.lock` to the end of `gemfile` entry.
```
Comment (colby-swandale) [inline bundler/lib/bundler/man/bundle-config.1.ronn]:
```suggestion
   The path to the lockfile that bundler should use. By default, Bundler adds
   `.lock` to the end of the `gemfile` entry. Can be set to `false` in the
   Gemfile to disable lockfile creation entirely (see gemfile(5)).
```

File: bundler/lib/bundler/man/bundle-config.1.ronn (hunk 6 above)
Comment (colby-swandale) [inline bundler/lib/bundler/man/bundle-config.1.ronn]: We have already updated the method in the Gemfile to be `lockfile`, and I think we should do that here so that we have a constant name used across this PR.

File: bundler/lib/bundler/man/bundle-install.1.ronn [hunk 8]
```diff
@@ -71,6 +71,12 @@ update process below under [CONSERVATIVE UPDATING][].
   does not remove any gems in the cache but keeps the newly bundled gems from
   being cached during the install.
 
+* `--no-lock`:
+  Do not create a lockfile. Useful if you want to install dependencies but not
+  lock versions of gems. Recommended for library development, and other
+  situations where the code is expected to work with a range of dependency
+  versions.
+
```
Comment (colby-swandale) [inline bundler/lib/bundler/man/bundle-install.1.ronn:80]:
```suggestion

  This has the same effect as using `lockfile false` in the Gemfile.
  See gemfile(5) for more information.
```

File: bundler/lib/bundler/man/gemfile.5.ronn [hunk 9]
```diff
@@ -556,3 +556,20 @@ bundler uses the following priority order:
...
+
+This is useful when you want to use different lockfiles per ruby version or
+platform.
+
+To avoid writing a lock file, use `false` as the argument:
+
+    lockfile false
+
+This is useful for library development and other situations where the code is
+expected to work with a range of dependency versions.
```
Comment (colby-swandale) [inline bundler/lib/bundler/man/gemfile.5.ronn:575]: Would you be able to add a section that's able to clarify the precedence. I'll need to double check but I believe it's:

  1. `--no-lock` CLI flag (disables lockfile)
  2. `lockfile` method in Gemfile
  3. `BUNDLE_LOCKFILE` environment variable
  4. Default behavior (Gemfile path + `.lock`)

File: bundler/lib/bundler/shared_helpers.rb [hunk 10]
```diff
@@ -297,6 +300,7 @@ def set_env(key, value)
     def set_bundle_variables
       Bundler::SharedHelpers.set_env "BUNDLE_BIN_PATH", bundle_bin_path
       Bundler::SharedHelpers.set_env "BUNDLE_GEMFILE", find_gemfile.to_s
+      Bundler::SharedHelpers.set_env "BUNDLE_LOCKFILE", default_lockfile.to_s
```
Comment (colby-swandale) [inline bundler/lib/bundler/shared_helpers.rb:303]: _Should_ this always be set for child processors? Or should we only set this when a non-default `Gemfile.lock` option is given to Bundler?

File: bundler/lib/bundler/man/bundle-config.1.ronn (hunk 6 above)
Comment (jeremyevans) [reply bundler/lib/bundler/man/bundle-config.1.ronn]: No, it wasn't intentional. And I assume we should rename the setting to `lockfile`. I'll make the change and add a spec for it. Thank you for catching this.

File: bundler/lib/bundler/shared_helpers.rb (hunk 10 above)
Comment (jeremyevans) [reply bundler/lib/bundler/shared_helpers.rb:303]: I thought it would be simpler to always set it. Making it conditional would make the code more complex. Do you think setting it would cause issues in some cases?

File: bundler/lib/bundler/man/gemfile.5.ronn (hunk 9 above)
Comment (jeremyevans) [reply bundler/lib/bundler/man/gemfile.5.ronn:575]: I believe that is the precedence as well. I can add a section for that.

**Comment by jeremyevans** [comment]
@colby-swandale Thank you for the review. I've updated the PR to incorporate all the changes you mentioned, except for the `shared_helpers.rb` change. I'm open to that change if you think it's necessary.

**Comment by kou** [comment]
Could you also update the PR title and description to use `lockfile` not `lock_file`?

**Comment by jeremyevans** [comment]
> Could you also update the PR title and description to use `lockfile` not `lock_file`?

Done.

File: bundler/lib/bundler/shared_helpers.rb (hunk 10 above)
Comment (colby-swandale) [reply bundler/lib/bundler/shared_helpers.rb:303]: I've given this question thought over the weekend and concluded  this the best solution going forward, it improves the consistency with the already present `BUNDLE_GEMFILE`.

File: bundler/spec/commands/install_spec.rb [hunk 11]
```diff
@@ -29,6 +29,28 @@
       expect(bundled_app_lock).to exist
     end
 
+    it "creates lockfile based on the lockfile method in Gemfile" do
```
Comment (colby-swandale) [inline bundler/spec/commands/install_spec.rb:32]: Should we make a warning when the user has a `Gemfile` & `Gemfile.lock`, but runs `bundle install` with `--no-lock` ?

File: bundler/lib/bundler/cli.rb [hunk 12]
```diff
@@ -219,6 +219,7 @@ def remove(*gems)
     method_option "local", type: :boolean, banner: "Do not attempt to fetch gems remotely and use the gem cache instead"
     method_option "prefer-local", type: :boolean, banner: "Only attempt to fetch gems remotely if not present locally, even if newer versions are available remotely"
     method_option "no-cache", type: :boolean, banner: "Don't update the existing gem cache."
+    method_option "no-lock", type: :boolean, banner: "Don't create a lockfile."
```
Comment (colby-swandale) [inline bundler/lib/bundler/cli.rb:222]: I'm curious, would people have a desire to see a CLI option to expose a `--lockfile` option? Similar to how we expose a `--gemfile` option. This might be a good "nice to have" for a follow-up PR rather than blocking this one.

File: bundler/spec/commands/install_spec.rb (hunk 11 above)
Comment (colby-swandale) [reply bundler/spec/commands/install_spec.rb:32]: Actually I've also found that when you run `--no-lock` with an existing `Gemfile` & `Gemfile.lock`, Bundler will load the existing `Gemfile.lock`. I think this _shouldn't_ be happening in this situation, but I'm keen to hear thoughts?

Comment (jeremyevans) [reply bundler/spec/commands/install_spec.rb:32]: As designed and documented, `--no-lock` means `Don't create a lockfile.`. Maybe it should be changed to `Don't write a lockfile.`, since it also prevents updating an existing lockfile? It doesn't affect whether an existing lockfile is read.

I think if we want to support ignoring an existing lockfile, we should have a separate option for that (e.g. `--ignore-lockfile` or `--ignore-existing-lockfile`). That option would prevent reading an existing lockfile, but wouldn't change whether a lock file is written. I think the two ideas are independent, one affects whether an existing lockfile is read, the other whether to write a lockfile.

File: bundler/lib/bundler/cli.rb (hunk 12 above)
Comment (jeremyevans) [reply bundler/lib/bundler/cli.rb:222]: That could be useful. But I agree it may be better to do in a follow-up PR.

**Comment by hsbt** [comment]
@colby-swandale @jeremyevans I would like to introduce this feature at 4.0.0. Could you create an issue or pull request if there are any additional concerns? Thanks!
